File: libjs/wModel/attributes.js
"use strict";

var ModelVocabulary = require("./vocabulary");
var Vocabulary = require("../wType/vocabulary");
var String = require("../wType/string");
var Uint64 = require("../wType/uint64");

var Attributes = ModelVocabulary.inherit({
    "className": "Attributes",
    "constructor": function(addr) {
        ModelVocabulary.fn.constructor.call(this, addr);
        
        this._attributes = global.Object.create(null);
    },
    "addAttribute": function(key, model) {
        var old = this._attributes[key];
        if (old) {
            throw new Error("Attribute with key " + key + " already exists");
        }
        this._attributes[key] = model;
        this.addModel(model);
        
        var vc = new Vocabulary();
        vc.insert("name", new String(key));
        vc.insert("address", model.getAddress());
        vc.insert("type", new Uint64(model.getType()));
        this.insert(key, vc);
    },
    "removeAttribute": function(key) {
        var model = this._attributes[key];
        if (!model) {
            throw new Error("An attempt to access non existing Attribute");
        }
        delete this._attributes[key];
        this.erase(key);
        this.removeModel(model);
        model.destructor();
    },
    "setAttribute": function(key, value) {
        this._attributes[key].set(value);
    }
});

module.exports = Attributes;

File: lib/wModel/attributes.cpp
#include "attributes.h"

M::Attributes::Attributes(const W::Address p_address, QObject* parent):
    M::Vocabulary(p_address, parent),
    attributes(new Map())
{
}

M::Attributes::~Attributes()
{
    delete attributes;
}

M::Model::ModelType M::Attributes::getType() const
{
    return type;
}

void M::Attributes::addAttribute(const W::String& key, M::Model* model)
{
    Map::const_iterator itr = attributes->find(key);
    if (itr != attributes->end()) {
        throw 1;
    }
    
    attributes->insert(std::make_pair(key, model));
    addModel(model);
    W::Vocabulary* vc = new W::Vocabulary();
    vc->insert(u"name", key);
    vc->insert(u"address", model->getAddress());
    vc->insert(u"type", W::Uint64(model->getType()));
    
    insert(key, vc);
}

void M::Attributes::removeAttribute(const W::String& key)
{
    Map::const_iterator itr = attributes->find(key);
    if (itr == attributes->end()) {
        throw 1;
    }
    M::Model* model = itr->second;
    attributes->erase(itr);
    erase(key);
    removeModel(model);
    
    delete model;
}

void M::Attributes::setAttribute(const W::String& key, const W::Object& value)
{
    Map::const_iterator itr = attributes->find(key);
    itr->second->set(value);
}

void M::Attributes::setAttribute(const W::String& key, W::Object* value)
{
    Map::const_iterator itr = attributes->find(key);
    itr->second->set(value);
}
